[Hinty] Add type hints to Scapy

**Project "Hinty"** aims at adding **Type hints** to Scapy. It will help discover bugs, improve the API, and make Scapy up-to-date with the high standards of Python libraries.

### Implementation

We use `mypy` to ensure automatic testing of the work that has already been completed. PRs that fall under project Hinty will process one (or a few) files and register them into the checks. The file `.config/mypy/mypy_enabled.txt` lists the files in which mypy checks are enabled. This process has been added by secdev/scapy#2162

Because we support Python 2.7, the format of the Type hints should be the following:
https://mypy.readthedocs.io/en/latest/cheat_sheet.html

## Guide on how to contribute

0. :cd: Make sure you've "git clonned" the latest version of scapy and are working from within it !

1. :pencil2: Pick a file you want to work on. If you're new on scapy, pick a file from within the `layers/` or `contrib/` folder that is small in size and complexity: at this point in time, the remaining layers in Scapy's core are often hard to type.

2. :pencil2: Add that file to `.config/mypy/mypy_enabled.txt`. This will enable it in our test suite.

3. ➡️ `pip install pyannotate cryptography tox mock` (install tox, pyannotate and cryptography)

4. ➡️ Run the tests for your file: `tox --sitepackages -- -x -K tcpdump -K manufdb -K wireshark -K tshark -K brotli -K zstd -K ci_only -K not_pyannotate -N`
You shall specify a `-e py[38]-[linux]_non_root` (fill the brackets with your OS infos) option before the `--sitepackages` option if you have multiple versions of python installed. To see all available tox tags (the whole `py...non_root` string), type `tox -l`. The `-x` option tells our test suite to use `pyannotate`.
This will create `test/pyannotate_results`
    - ⚠️ ✏️ In many cases, pyannotate will glitch out (https://github.com/dropbox/pyannotate/issues/92). You need to open `pyannotate_results` with your favorite editor and replace all occurrences of `<locals>` (with the <>) by blanks (or anything really). For instance on vim: `:%s/<locals>//g`

5. ➡️ `pyannotate --type-info test/pyannotate_results -w [the file.py you are working on] `

6. ✏️ The file has been automatically processed. Now, edit it to fix the mistakes, and check your work with `tox -e mypy`. This might take time
    - ℹ️ Remember that you can use `cast(type, obj)` from `typing` to help mypy without affecting the code.
    - ⚠️ **Import any types you might need from `scapy.compat` instead of `typing`**. It provides a fallback if `typing` isn't installed (python < 3.5). I know it's annoying.
    - ℹ️ If a class extends `Field` directly (very few do) it will require to be passed the internal and machine representation of the field. See [this](https://scapy.readthedocs.io/en/latest/build_dissect.html#manipulating-packets-manipulating-its-fields) if you don't know what I'm talking about. For instance: `class subField(Field[float, int]):` is a field where `i=float` and `m=int`. In most cases, it should be `[int, int]` (a normal integer field).

7. ⚠️ If you get `error: No library stub file for module ...`, add an exception for it in `.config/mypy/mypy.ini` (follow the existing format)

8. ✅ Check that the files still pass PEP8: `tox -e flake8`

9. 🥇 When everything passes, submit a PR. :smile:

## Active hinty related issues

- Discuss strict equality in Mypy https://github.com/secdev/scapy/issues/2823

### Documentation for advanced typing (that I always lose the links for..)

- [Generics](https://mypy.readthedocs.io/en/stable/generics.html)
- [Methods overloading](https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading)

## Hinty status

**MyPy Support: 25.44%**
| Module | Typed code (lines) | Typed files |
| --- | --- | --- |
|`[core]` | 100.00% | 33/33 |
|`arch` | 100.00% | 13/13 |
|`layers` | 3.06% | 2/92 |
|`contrib` | 14.28% | 27/167 |
|`libs` | 23.30% | 6/8 |
|`asn1` | 100.00% | 4/4 |
|`tools` | 11.08% | 1/9 |
|`modules` | 0.00% | 0/8 |

Ok, so I tried adding plist.py(after removing utils.py) to the mypy_enabled file, and I get the same errors on the main.py and http2.py file
`root@kali:~/scapy1/scapy# tox -e mypy
mypy installed: mypy==0.750,mypy-extensions==0.4.3,typed-ast==1.4.0,typing==3.7.4.1,typing-extensions==3.7.4.1
mypy run-test-pre: PYTHONHASHSEED='4222907011'
mypy run-test: commands[0] | python .config/mypy/mypy_check.py
scapy/main.py:159: error: Module has no attribute "iteritems"
scapy/main.py:395: error: Module has no attribute "iteritems"
scapy/plist.py:28: error: Cannot find implementation or library stub for module named 'scapy.modules.six.moves'
scapy/plist.py:28: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
scapy/plist.py:302: error: Module has no attribute "iteritems"
scapy/plist.py:305: error: Module has no attribute "iteritems"
scapy/contrib/http2.py:110: error: Module has no attribute "integer_types"
scapy/contrib/http2.py:202: error: Module has no attribute "integer_types"
scapy/contrib/http2.py:220: error: Module has no attribute "integer_types"
scapy/contrib/http2.py:346: error: Module has no attribute "integer_types"
scapy/contrib/http2.py:349: error: Module has no attribute "integer_types"
scapy/contrib/http2.py:353: error: Module has no attribute "integer_types"
scapy/contrib/http2.py:2071: error: Module has no attribute "integer_types"
scapy/contrib/http2.py:2429: error: Module has no attribute "iteritems"
scapy/contrib/http2.py:2448: error: Module has no attribute "iteritems"
Found 18 errors in 3 files (checked 4 source files)
ERROR: InvocationError for command /root/scapy1/scapy/.tox/mypy/bin/python .config/mypy/mypy_check.py (exited with code 1)
`

All of these are functions in the six.py module. And I can't figure out how to fix them.

--- a/.config/mypy/mypy.ini
+++ b/.config/mypy/mypy.ini
@@ -12,3 +12,9 @@ ignore_missing_imports = True
 [mypy-scapy.modules.six,scapy.modules.six.moves,scapy.libs.winpcapy]
 ignore_errors = True
 ignore_missing_imports = True
+
+[mypy-pyx]
+ignore_missing_imports = True
+
+[mypy-matplotlib.lines]
+ignore_missing_imports = True

--- a/.config/mypy/mypy.ini
+++ b/.config/mypy/mypy.ini
@@ -18,3 +18,4 @@ ignore_missing_imports = True
 
 [mypy-matplotlib.lines]
 ignore_missing_imports = True
+

--- a/.config/mypy/mypy_enabled.txt
+++ b/.config/mypy/mypy_enabled.txt
@@ -7,3 +7,4 @@
 scapy/__init__.py
 scapy/main.py
 scapy/contrib/http2.py
+scapy/plist.py

--- a/scapy/extlib.py
+++ b/scapy/extlib.py
@@ -20,6 +20,7 @@ from scapy.error import log_loading
 try:
     from matplotlib import get_backend as matplotlib_get_backend
     from matplotlib import pyplot as plt
+    from matplotlib.lines import Line2D
     MATPLOTLIB = 1
     if "inline" in matplotlib_get_backend():
         MATPLOTLIB_INLINED = 1
@@ -29,6 +30,7 @@ try:
 # RuntimeError to catch gtk "Cannot open display" error
 except (ImportError, RuntimeError):
     plt = None
+    Line2D = None
     MATPLOTLIB = 0
     MATPLOTLIB_INLINED = 0
     MATPLOTLIB_DEFAULT_PLOT_KARGS = dict()

--- a/scapy/plist.py
+++ b/scapy/plist.py
@@ -13,6 +13,7 @@ from __future__ import print_function
 import os
 from collections import defaultdict
 
+
 from scapy.compat import lambda_tuple_converter
 from scapy.config import conf
 from scapy.base_classes import BasePacket, BasePacketList, _CanvasDumpExtended
@@ -21,8 +22,10 @@ from scapy.utils import do_graph, hexdump, make_table, make_lined_table, \
 from scapy.extlib import plt, MATPLOTLIB_INLINED, MATPLOTLIB_DEFAULT_PLOT_KARGS
 from functools import reduce
 import scapy.modules.six as six
+from matplotlib.lines import Line2D
 from scapy.modules.six.moves import range, zip
-
+from scapy.compat import Optional, List, Union, Tuple, Dict, Any, Callable
+from scapy.packet import Packet
 #############
 #  Results  #
 #############
@@ -46,18 +49,23 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         self.listname = name
 
     def __len__(self):
+        # type: () -> int
         return len(self.res)
 
     def _elt2pkt(self, elt):
+        # type: (Packet) -> Packet
         return elt
 
     def _elt2sum(self, elt):
+        # type: (Packet) -> str
         return elt.summary()
 
     def _elt2show(self, elt):
+        # type: (Packet) -> str
         return self._elt2sum(elt)
 
     def __repr__(self):
+        # type: () -> str
         stats = {x: 0 for x in self.stats}
         other = 0
         for r in self.res:
@@ -85,6 +93,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
                                ct.punct(">"))
 
     def __getstate__(self):
+        # type: () -> Dict[str, Any]
         """
         Creates a basic representation of the instance, used in
         conjunction with __setstate__() e.g. by pickle
@@ -99,6 +108,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return state
 
     def __setstate__(self, state):
+        # type: (Dict[str, Any]) -> None
         """
         Sets instance attributes to values given by state, used in
         conjunction with __getstate__() e.g. by pickle
@@ -110,6 +120,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         self.listname = state['listname']
 
     def __getattr__(self, attr):
+        # type: (str) -> Any
         return getattr(self.res, attr)
 
     def __getitem__(self, item):
@@ -126,10 +137,12 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
                               name="mod %s" % self.listname)
 
     def __add__(self, other):
+        # type: (PacketList) -> PacketList
         return self.__class__(self.res + other.res,
                               name="%s+%s" % (self.listname, other.listname))
 
     def summary(self, prn=None, lfilter=None):
+        # type: (Optional[Callable], Optional[Callable]) -> None
         """prints a summary of each packet
 
         :param prn: function to apply to each packet instead of
@@ -147,6 +160,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
                 print(prn(r))
 
     def nsummary(self, prn=None, lfilter=None):
+        # type: (Optional[Callable], Optional[Callable]) -> None
         """prints a summary of each packet with the packet's number
 
         :param prn: function to apply to each packet instead of
@@ -169,29 +183,35 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         self.show()
 
     def show(self, *args, **kargs):
+        # type: (Any, Any) -> None
         """Best way to display the packet list. Defaults to nsummary() method"""  # noqa: E501
         return self.nsummary(*args, **kargs)
 
     def filter(self, func):
+        # type: (Callable) -> PacketList
         """Returns a packet list filtered by a truth function. This truth
         function has to take a packet as the only argument and return a boolean value."""  # noqa: E501
         return self.__class__([x for x in self.res if func(x)],
                               name="filtered %s" % self.listname)
 
     def make_table(self, *args, **kargs):
+        # type: (Any, Any) -> None
         """Prints a table using a function that returns for each packet its head column value, head row value and displayed value  # noqa: E501
         ex: p.make_table(lambda x:(x[IP].dst, x[TCP].dport, x[TCP].sprintf("%flags%")) """  # noqa: E501
         return make_table(self.res, *args, **kargs)
 
     def make_lined_table(self, *args, **kargs):
+        # type: (Any, Any) -> None
         """Same as make_table, but print a table with lines"""
         return make_lined_table(self.res, *args, **kargs)
 
     def make_tex_table(self, *args, **kargs):
+        # type: (Any, Any) -> None
         """Same as make_table, but print a table with LaTeX syntax"""
         return make_tex_table(self.res, *args, **kargs)
 
     def plot(self, f, lfilter=None, plot_xy=False, **kargs):
+        # type: (Callable, Optional[Callable], bool, Any) -> Line2D
         """Applies a function to each packet to get a value that will be plotted
         with matplotlib. A list of matplotlib.lines.Line2D is returned.
 
@@ -223,6 +243,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return lines
 
     def diffplot(self, f, delay=1, lfilter=None, **kargs):
+        # type: (Callable, int, Optional[Callable], Any) -> Line2D
         """diffplot(f, delay=1, lfilter=None)
         Applies a function to couples (l[i],l[i+delay])
 
@@ -250,6 +271,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return lines
 
     def multiplot(self, f, lfilter=None, plot_xy=False, **kargs):
+        # type: (Callable, Optional[Callable], bool, Any) -> Line2D
         """Uses a function that returns a label and a value for this label, then
         plots all the values label by label.
 
@@ -267,7 +289,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
             lst_pkts = (f(*e) for e in self.res if lfilter(*e))
 
         # Apply the function f to the packets
-        d = {}
+        d = {}  # type: Dict[str, List[float]]
         for k, v in lst_pkts:
             d.setdefault(k, []).append(v)
 
@@ -290,11 +312,13 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return lines
 
     def rawhexdump(self):
+        # type: (Optional[Callable]) -> None
         """Prints an hexadecimal dump of each packet in the list"""
         for p in self:
             hexdump(self._elt2pkt(p))
 
     def hexraw(self, lfilter=None):
+        # type: (Optional[Callable]) -> None
         """Same as nsummary(), except that if a packet has a Raw layer, it will be hexdumped  # noqa: E501
         lfilter: a truth function that decides whether a packet must be displayed"""  # noqa: E501
         for i, res in enumerate(self.res):
@@ -308,6 +332,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
                 hexdump(p.getlayer(conf.raw_layer).load)
 
     def hexdump(self, lfilter=None):
+        # type: (Optional[Callable]) -> None
         """Same as nsummary(), except that packets are also hexdumped
         lfilter: a truth function that decides whether a packet must be displayed"""  # noqa: E501
         for i, res in enumerate(self.res):
@@ -320,6 +345,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
             hexdump(p)
 
     def padding(self, lfilter=None):
+        # type: (Optional[Callable]) -> None
         """Same as hexraw(), for Padding layer"""
         for i, res in enumerate(self.res):
             p = self._elt2pkt(res)
@@ -331,6 +357,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
                     hexdump(p.getlayer(conf.padding_layer).load)
 
     def nzpadding(self, lfilter=None):
+        # type: (Optional[Callable]) -> None
         """Same as padding() but only non null padding"""
         for i, res in enumerate(self.res):
             p = self._elt2pkt(res)
@@ -393,6 +420,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return do_graph(gr, **kargs)
 
     def afterglow(self, src=None, event=None, dst=None, **kargs):
+        # type: (Optional[Callable], Optional[Callable], Optional[Callable], Any) -> None  # noqa: E501
         """Experimental clone attempt of http://sourceforge.net/projects/afterglow
         each datum is reduced as src -> event -> dst and the data are graphed.
         by default we have IP.src -> IP.dport -> IP.dst"""
@@ -402,9 +430,9 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
             event = lambda x: x['IP'].dport
         if dst is None:
             dst = lambda x: x['IP'].dst
-        sl = {}
-        el = {}
-        dl = {}
+        sl = {}  # type: Dict[Any, Tuple[int, Any]]
+        el = {}  # type: Dict[Any, Tuple[int, Any]]
+        dl = {}  # type: Dict[Any, Tuple[int, Any]]
         for i in self.res:
             try:
                 s, e, d = src(i), event(i), dst(i)
@@ -472,6 +500,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return do_graph(gr, **kargs)
 
     def canvas_dump(self, **kargs):
+        # type: (Any) -> Any  # Using Any since pyx is imported later
         import pyx
         d = pyx.document.document()
         len_res = len(self.res)
@@ -487,6 +516,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return d
 
     def sr(self, multi=0):
+        # type: (int) -> Tuple[SndRcvList, PacketList]
         """sr([multi=1]) -> (SndRcvList, PacketList)
         Matches packets in the list and return ( (matched couples), (unmatched packets) )"""  # noqa: E501
         remain = self.res[:]
@@ -549,6 +579,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return dict(sessions)
 
     def replace(self, *args, **kargs):
+        # type: (Any, Any) -> PacketList
         """
         lst.replace(<field>,[<oldvalue>,]<newvalue>)
         lst.replace( (fld,[ov],nv),(fld,[ov,]nv),...)
@@ -581,7 +612,13 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
             x.append(p)
         return x
 
-    def getlayer(self, cls, nb=None, flt=None, name=None, stats=None):
+    def getlayer(self, cls,  # type: Packet
+                 nb=None,  # type: Optional[int]
+                 flt=None,  # type: Optional[Dict[str, Any]]
+                 name=None,  # type: Optional[str]
+                 stats=None  # type: Optional[List[Packet]]
+                 ):
+        # type: (...) -> PacketList
         """Returns the packet list from a given layer.
 
         See ``Packet.getlayer`` for more info.
@@ -608,7 +645,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         if stats is None:
             stats = self.stats
 
-        getlayer_arg = {}
+        getlayer_arg = {}  # type: Dict[str, Any]
         if flt is not None:
             getlayer_arg.update(flt)
         getlayer_arg['cls'] = cls
@@ -623,6 +660,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         )
 
     def convert_to(self, other_cls, name=None, stats=None):
+        # type: (Packet, Optional[str], Optional[List[Packet]]) -> PacketList
         """Converts all packets to another type.
 
         See ``Packet.convert_to`` for more info.
@@ -652,13 +690,16 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
 
 
 class SndRcvList(PacketList):
-    __slots__ = []
+    __slots__ = []  # type: List[Any]
 
     def __init__(self, res=None, name="Results", stats=None):
+        # type: (Optional[Union[List[Packet], PacketList]], str, Optional[List[Packet]]) -> None  # noqa: E501
         PacketList.__init__(self, res, name, stats)
 
     def _elt2pkt(self, elt):
+        # type: (Tuple[Packet, Packet]) -> Packet
         return elt[1]
 
     def _elt2sum(self, elt):
+        # type: (Tuple[Packet, Packet]) -> str
         return "%s ==> %s" % (elt[0].summary(), elt[1].summary())

--- a/scapy/plist.py
+++ b/scapy/plist.py
@@ -17,9 +17,10 @@ from collections import defaultdict
 from scapy.compat import lambda_tuple_converter
 from scapy.config import conf
 from scapy.base_classes import BasePacket, BasePacketList, _CanvasDumpExtended
+from scapy.fields import IPField, ShortEnumField, PacketField
 from scapy.utils import do_graph, hexdump, make_table, make_lined_table, \
     make_tex_table, issubtype
-from scapy.extlib import plt, MATPLOTLIB, \
+from scapy.extlib import plt, Line2D, \
     MATPLOTLIB_INLINED, MATPLOTLIB_DEFAULT_PLOT_KARGS
 from functools import reduce
 import scapy.modules.six as six
@@ -93,7 +94,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
                                ct.punct(">"))
 
     def __getstate__(self):
-        # type: () -> Dict[str, Any]
+        # type: () -> Dict[str, Union[List[PacketField], List[Packet], str]]
         """
         Creates a basic representation of the instance, used in
         conjunction with __setstate__() e.g. by pickle
@@ -108,7 +109,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return state
 
     def __setstate__(self, state):
-        # type: (Dict[str, Any]) -> None
+        # type: (Dict[str, Union[List[PacketField], List[Packet], str]]) -> None  # noqa: E501
         """
         Sets instance attributes to values given by state, used in
         conjunction with __getstate__() e.g. by pickle
@@ -211,7 +212,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return make_tex_table(self.res, *args, **kargs)
 
     def plot(self, f, lfilter=None, plot_xy=False, **kargs):
-        # type: (Callable, Optional[Callable], bool, Any) -> MATPLOTLIB.lines.Line2D  # noqa: E501
+        # type: (Callable, Optional[Callable], bool, Any) -> Line2D
         """Applies a function to each packet to get a value that will be plotted
         with matplotlib. A list of matplotlib.lines.Line2D is returned.
 
@@ -243,7 +244,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return lines
 
     def diffplot(self, f, delay=1, lfilter=None, **kargs):
-        # type: (Callable, int, Optional[Callable], Any) -> MATPLOTLIB.lines.Line2D  # noqa: E501
+        # type: (Callable, int, Optional[Callable], Any) -> Line2D
         """diffplot(f, delay=1, lfilter=None)
         Applies a function to couples (l[i],l[i+delay])
 
@@ -271,7 +272,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return lines
 
     def multiplot(self, f, lfilter=None, plot_xy=False, **kargs):
-        # type: (Callable, Optional[Callable], bool, Any) -> MATPLOTLIB.lines.Line2D  # noqa: E501
+        # type: (Callable, Optional[Callable], bool, Any) -> Line2D
         """Uses a function that returns a label and a value for this label, then
         plots all the values label by label.
 
@@ -430,9 +431,9 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
             event = lambda x: x['IP'].dport
         if dst is None:
             dst = lambda x: x['IP'].dst
-        sl = {}  # type: Dict[Any, Tuple[int, Any]]
-        el = {}  # type: Dict[Any, Tuple[int, Any]]
-        dl = {}  # type: Dict[Any, Tuple[int, Any]]
+        sl = {}  # type: Dict[IPField, Tuple[int, List[ShortEnumField]]]
+        el = {}  # type: Dict[ShortEnumField, Tuple[int, List[IPField]]]
+        dl = {}  # type: Dict[IPField, ShortEnumField]
         for i in self.res:
             try:
                 s, e, d = src(i), event(i), dst(i)
@@ -690,10 +691,14 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
 
 
 class SndRcvList(PacketList):
-    __slots__ = []  # type: List[Any]
+    __slots__ = []  # type: List[str]
 
-    def __init__(self, res=None, name="Results", stats=None):
-        # type: (Optional[Union[List[Packet], PacketList]], str, Optional[List[Packet]]) -> None  # noqa: E501
+    def __init__(self,
+                 res=None,  # type: Optional[Union[List[Packet], PacketList]]
+                 name="Results",  # type: str
+                 stats=None  # type: Optional[List[Packet]]
+                 ):
+        # type: (...) -> None
         PacketList.__init__(self, res, name, stats)
 
     def _elt2pkt(self, elt):

--- a/scapy/plist.py
+++ b/scapy/plist.py
@@ -13,7 +13,6 @@ from __future__ import print_function
 import os
 from collections import defaultdict
 
-
 from scapy.compat import lambda_tuple_converter
 from scapy.config import conf
 from scapy.base_classes import BasePacket, BasePacketList, _CanvasDumpExtended

--- a/scapy/plist.py
+++ b/scapy/plist.py
@@ -19,10 +19,10 @@ from scapy.config import conf
 from scapy.base_classes import BasePacket, BasePacketList, _CanvasDumpExtended
 from scapy.utils import do_graph, hexdump, make_table, make_lined_table, \
     make_tex_table, issubtype
-from scapy.extlib import plt, MATPLOTLIB_INLINED, MATPLOTLIB_DEFAULT_PLOT_KARGS
+from scapy.extlib import plt, MATPLOTLIB, \
+    MATPLOTLIB_INLINED, MATPLOTLIB_DEFAULT_PLOT_KARGS
 from functools import reduce
 import scapy.modules.six as six
-from matplotlib.lines import Line2D
 from scapy.modules.six.moves import range, zip
 from scapy.compat import Optional, List, Union, Tuple, Dict, Any, Callable
 from scapy.packet import Packet
@@ -211,7 +211,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return make_tex_table(self.res, *args, **kargs)
 
     def plot(self, f, lfilter=None, plot_xy=False, **kargs):
-        # type: (Callable, Optional[Callable], bool, Any) -> Line2D
+        # type: (Callable, Optional[Callable], bool, Any) -> MATPLOTLIB.lines.Line2D  # noqa: E501
         """Applies a function to each packet to get a value that will be plotted
         with matplotlib. A list of matplotlib.lines.Line2D is returned.
 
@@ -243,7 +243,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return lines
 
     def diffplot(self, f, delay=1, lfilter=None, **kargs):
-        # type: (Callable, int, Optional[Callable], Any) -> Line2D
+        # type: (Callable, int, Optional[Callable], Any) -> MATPLOTLIB.lines.Line2D  # noqa: E501
         """diffplot(f, delay=1, lfilter=None)
         Applies a function to couples (l[i],l[i+delay])
 
@@ -271,7 +271,7 @@ class PacketList(BasePacketList, _CanvasDumpExtended):
         return lines
 
     def multiplot(self, f, lfilter=None, plot_xy=False, **kargs):
-        # type: (Callable, Optional[Callable], bool, Any) -> Line2D
+        # type: (Callable, Optional[Callable], bool, Any) -> MATPLOTLIB.lines.Line2D  # noqa: E501
         """Uses a function that returns a label and a value for this label, then
         plots all the values label by label.
 

--- a/test/regression.uts
+++ b/test/regression.uts
@@ -680,18 +680,18 @@ from mock import MagicMock, patch
 bck_scapy_ext_lib = sys.modules.get("scapy.extlib", None)
 del(sys.modules["scapy.extlib"])
 
-matplotlib = MagicMock()
-matplotlib.get_backend.return_value = "inline"
-matplotlib.pyplot = MagicMock()
-matplotlib.pyplot.plt = None
-with patch.dict("sys.modules", matplotlib=matplotlib):
-    from scapy.extlib import MATPLOTLIB, MATPLOTLIB_INLINED, MATPLOTLIB_DEFAULT_PLOT_KARGS
+mock_matplotlib = MagicMock()
+mock_matplotlib.get_backend.return_value = "inline"
+mock_matplotlib.pyplot = MagicMock()
+mock_matplotlib.pyplot.plt = None
+with patch.dict("sys.modules", **{ "matplotlib": mock_matplotlib, "matplotlib.lines": mock_matplotlib}):
+    from scapy.extlib import MATPLOTLIB, MATPLOTLIB_INLINED, MATPLOTLIB_DEFAULT_PLOT_KARGS, Line2D
     assert MATPLOTLIB == 1
     assert MATPLOTLIB_INLINED == 1
     assert "marker" in MATPLOTLIB_DEFAULT_PLOT_KARGS
 
-matplotlib.get_backend.return_value = "ko"
-with patch.dict("sys.modules", matplotlib=matplotlib):
+mock_matplotlib.get_backend.return_value = "ko"
+with patch.dict("sys.modules", **{ "matplotlib": mock_matplotlib, "matplotlib.lines": mock_matplotlib}):
     from scapy.extlib import MATPLOTLIB, MATPLOTLIB_INLINED, MATPLOTLIB_DEFAULT_PLOT_KARGS
     assert MATPLOTLIB == 1
     assert MATPLOTLIB_INLINED == 0

